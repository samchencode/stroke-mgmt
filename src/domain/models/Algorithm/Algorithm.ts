import type { AlgorithmId } from '@/domain/models/Algorithm/AlgorithmId';
import type { Outcome } from '@/domain/models/Algorithm/Outcome';

interface Algorithm {
  getId(): AlgorithmId;
  getTitle(): string;
  getBody(): string;
  getOutcomes(): Outcome[];
  hasOutcome(): boolean;
  is(other: Algorithm): boolean;
}

export type { Algorithm };
