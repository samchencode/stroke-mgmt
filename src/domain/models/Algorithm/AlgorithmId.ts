class AlgorithmId {
  private id: string;

  constructor(id: string) {
    this.id = id;
  }

  getId() {
    return this.id;
  }

  is(v: AlgorithmId) {
    return this.id === v.getId();
  }

  toString() {
    return this.getId();
  }
}

export { AlgorithmId };
