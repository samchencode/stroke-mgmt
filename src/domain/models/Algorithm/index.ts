export type { Algorithm } from '@/domain/models/Algorithm/Algorithm';
export type { AlgorithmVisitor } from '@/domain/models/Algorithm/AlgorithmVisitor';
export { TextAlgorithm } from '@/domain/models/Algorithm/TextAlgorithm';
export { ScoredAlgorithm } from '@/domain/models/Algorithm/ScoredAlgorithm';
export { AlgorithmInfo } from '@/domain/models/Algorithm/AlgorithmInfo';
export { AlgorithmId } from '@/domain/models/Algorithm/AlgorithmId';
export { Outcome } from '@/domain/models/Algorithm/Outcome';
export type { AlgorithmRepository } from '@/domain/models/Algorithm/ports/AlgorithmRepository';
export type { AlgorithmRenderer } from '@/domain/models/Algorithm/ports/AlgorithmRenderer';
export {
  GreaterThanCriterion,
  LessThanCriterion,
} from '@/domain/models/Algorithm/Criterion';
export {
  Switch,
  YesNoSwitch,
  SwitchId,
} from '@/domain/models/Algorithm/Switch';
export { RenderedAlgorithm } from '@/domain/models/Algorithm/RenderedAlgorithm';
export { RenderedAlgorithmCollection } from '@/domain/models/Algorithm/RenderedAlgorithmCollection';
