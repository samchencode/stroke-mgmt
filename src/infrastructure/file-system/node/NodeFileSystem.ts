import fs from 'fs';
import path from 'path';
import type { FileSystem } from '@/infrastructure/file-system/FileSystem';

const readUtf8File = (p: string) =>
  new Promise<string>((resolve, reject) => {
    fs.readFile(p, 'utf-8', (err, data) => (err ? reject(err) : resolve(data)));
  });

class NodeFileSystem implements FileSystem<'jest'> {
  async getAssetAsString(aliasedFilePath: string): Promise<string> {
    const filePath = aliasedFilePath.replace(
      /^@/,
      path.resolve(__dirname, '../../..')
    );
    return readUtf8File(filePath);
  }
}

export { NodeFileSystem };
