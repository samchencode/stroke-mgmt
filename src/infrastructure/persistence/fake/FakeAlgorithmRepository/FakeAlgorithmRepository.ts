import type {
  Algorithm,
  AlgorithmId,
  AlgorithmRepository,
} from '@/domain/models/Algorithm';
import { algorithms } from '@/infrastructure/persistence/fake/FakeAlgorithmRepository/fakeAlgorithms';

class FakeAlgorithmRepository implements AlgorithmRepository {
  async getAll(): Promise<Algorithm[]> {
    return algorithms;
  }

  async getAllShownOnHomeScreen(): Promise<Algorithm[]> {
    return algorithms;
  }

  async getById(id: AlgorithmId): Promise<Algorithm> {
    const result = algorithms.find((a) => a.getId().is(id));
    if (!result) throw Error('Algorithm not found!');
    return result;
  }
}

export { FakeAlgorithmRepository };
