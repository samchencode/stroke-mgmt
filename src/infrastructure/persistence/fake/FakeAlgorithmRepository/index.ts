export { FakeAlgorithmRepository } from '@/infrastructure/persistence/fake/FakeAlgorithmRepository/FakeAlgorithmRepository';
