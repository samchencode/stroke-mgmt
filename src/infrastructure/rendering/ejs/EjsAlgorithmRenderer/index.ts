export { EjsAlgorithmRenderer } from '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/EjsAlgorithmRenderer';
export type {
  Event,
  OutcomeSelectedEvent,
  ErrorEvent,
  LayoutEvent,
  SwitchChangedEvent,
} from '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/Event';
