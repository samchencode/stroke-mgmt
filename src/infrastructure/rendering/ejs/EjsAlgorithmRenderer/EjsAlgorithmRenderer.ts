import ejs from 'ejs';
import type {
  Algorithm,
  AlgorithmRenderer,
  Outcome,
} from '@/domain/models/Algorithm';
import type { FileSystem } from '@/infrastructure/file-system/FileSystem';
import { EjsAlgorithmVisitor } from '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/EjsAlgorithmVisitor';

class EjsAlgorithmRenderer implements AlgorithmRenderer {
  ready: Promise<void>;

  fs: FileSystem<unknown>;

  textAlgorithmTemplate: ejs.TemplateFunction | null = null;

  scoredAlgorithmTemplate: ejs.TemplateFunction | null = null;

  outcomeTemplate: ejs.TemplateFunction | null = null;

  constructor(fileSystem: FileSystem<unknown>) {
    this.fs = fileSystem;
    this.ready = this.init();
  }

  async init(): Promise<void> {
    const textAlgo = await this.fs.getAssetAsString(
      require('@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/textAlgorithm.ejs')
    );
    this.textAlgorithmTemplate = ejs.compile(textAlgo);
    const scoredAlgo = await this.fs.getAssetAsString(
      require('@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/scoredAlgorithm.ejs')
    );
    this.scoredAlgorithmTemplate = ejs.compile(scoredAlgo);
    const outcome = await this.fs.getAssetAsString(
      require('@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/outcome.ejs')
    );
    this.outcomeTemplate = ejs.compile(outcome);
  }

  async renderAlgorithm(algorithm: Algorithm): Promise<string> {
    await this.ready;
    if (!this.textAlgorithmTemplate || !this.scoredAlgorithmTemplate)
      throw Error('Algorithm templates not found!');

    const visitor = new EjsAlgorithmVisitor(
      this.textAlgorithmTemplate,
      this.scoredAlgorithmTemplate
    );
    visitor.visit(algorithm);
    return visitor.getRenderedHtml();
  }

  async renderOutcome(outcome: Outcome): Promise<string> {
    await this.ready;
    if (!this.outcomeTemplate) throw Error('Outcome template not found!');
    return this.outcomeTemplate({ outcome });
  }
}

export { EjsAlgorithmRenderer };
