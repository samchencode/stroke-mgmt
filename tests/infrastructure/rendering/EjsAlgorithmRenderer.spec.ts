import { AlgorithmId } from '@/domain/models/Algorithm';
import { NodeFileSystem } from '@/infrastructure/file-system/node/NodeFileSystem';
import { FakeAlgorithmRepository } from '@/infrastructure/persistence/fake/FakeAlgorithmRepository';
import { EjsAlgorithmRenderer } from '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer';

describe('EjsAlgorithmRenderer', () => {
  let repo: FakeAlgorithmRepository;
  let fs: NodeFileSystem;
  beforeAll(() => {
    jest.resetModules();
    jest.doMock(
      '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/textAlgorithm.ejs',
      () =>
        '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/textAlgorithm.ejs'
    );
    jest.doMock(
      '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/scoredAlgorithm.ejs',
      () =>
        '@/infrastructure/rendering/ejs/EjsAlgorithmRenderer/scoredAlgorithm.ejs'
    );
    fs = new NodeFileSystem();
    repo = new FakeAlgorithmRepository();
  });

  describe('Instantiation', () => {
    it('should be created given file system with template', () => {
      const create = () => new EjsAlgorithmRenderer(fs);
      expect(create).not.toThrowError();
    });
  });

  describe('Behavior', () => {
    it('should render text algorithm', async () => {
      const textAlgo = await repo.getById(new AlgorithmId('0'));
      const renderer = new EjsAlgorithmRenderer(fs);
      const result = await renderer.render(textAlgo);
      expect(result).toMatchSnapshot();
    });

    it('should render scored algorithm', async () => {
      const scoredAlgo = await repo.getById(new AlgorithmId('1'));
      const renderer = new EjsAlgorithmRenderer(fs);
      const result = await renderer.render(scoredAlgo);
      expect(result).toMatchSnapshot();
    });
  });
});
