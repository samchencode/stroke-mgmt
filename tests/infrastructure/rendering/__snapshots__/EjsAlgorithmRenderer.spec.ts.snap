// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EjsAlgorithmRenderer Behavior should render scored algorithm 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>GWNS Algorithm</title>
  <style>
    .template {
      box-sizing: border-box;
      margin: 0;
      padding: 0;
    }

    .template.switches__list,
    .template.outcomes__list {
      list-style: none;
      padding: 0;
    }

    .template.header__title {
      margin: 0;
      /* headline-medium */
      line-height: 36px;
      font-size: 28px;
      font-weight: 400;
    }

    .template.body__paragraph {
      margin-top: 16px;
      /* body-medium */
      line-height: 20px;
      font-size: 14px;
      letter-spacing: 0.25px;
      font-weight: 400;
    }

    .template.switches__list {
      margin-top: 16px;
    }

    .template.switches__list-item {
      display: flex;
      justify-content: space-between;
      align-items: center;
      padding-top: 8px;
      user-select: none;
    }

    .template.switches__list-item:first-child {
      padding-top: 0;
    }

    .template.switches__input {
      height: 0px;
      width: 0px;
      visibility: hidden;
    }

    .template.switches__group-label,
    .template.switches__radio-label {
      /* label-large */
      line-height: 20px;
      font-size: 14px;
      letter-spacing: 0.1px;
      font-weight: 500;
    }

    .template.switches__radiogroup {
      position: relative;
      padding: 4px 8px 4px 8px;
      border-radius: 999px;
      outline: 1px solid black;
      overflow: hidden;
    }

    .template.switches__radio-label {
      display: inline-block;
      width: 2em;
      text-align: center;
    }

    .template.switches__radiogroup-bg {
      background-color: black;
      position: absolute;
      top: 0;
      width: 50%;
      height: 100%;
      z-index: -1;
      display: none;
    }

    .template.switches__input:checked+.template.switches__radio-label {
      color: white;
    }

    .template.switches__input--yes:checked~.template.switches__radiogroup-bg {
      display: block;
      left: 0;
    }

    .template.switches__input--no:checked~.template.switches__radiogroup-bg {
      display: block;
      right: 0;
    }

    .template.outcomes__list-item {
      margin-top: 8px;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
    }

    .template.outcomes__list-item:first-child {
      margin-top: 8px;
    }

    .template.outcomes__button {
      background-color: black;
      color: white;
      border-radius: 999px;
      height: 40px;
      padding: 0 24px;
      /* label-large */
      line-height: 20px;
      font-size: 14px;
      letter-spacing: 0.1px;
      font-weight: 500;
    }
  </style>
</head>

<body>
  <header>
    <h1 class="template header__title">GWNS Algorithm</h1>
  </header>
  <main>
    <p class="template body__paragraph">
      All need immediate Rx and activate brain attack team immediately. ABCT = (airway, breathing, circulation, time of stroke onset). Endotracheal intubation if compromised airway or GCS less than eight. Maintain SBP 140 to 180 mmHg. HR 60-100. Establish IV access and hydration is recommended based on height/weight and clinical scenarios
    </p>
    <ul class="template switches__list">
      
      <li class="template switches__list-item">
        <span class="template switches__group-label">Gaze Deviation</span>
        <div class="template switches__radiogroup">
          <input class="template switches__input switches__input--yes" name="switch-0" id="switch-0-yes" type="radio">
          <label class="template switches__radio-label switches__radio-label--yes" for="switch-0-yes">Yes</label>
          <input class="template switches__input switches__input--no" name="switch-0" id="switch-0-no" type="radio">
          <label class="template switches__radio-label switches__radio-label--no" for="switch-0-no">No</label>
          <div class="template switches__radiogroup-bg"></div>
        </div>
      </li>
      
      <li class="template switches__list-item">
        <span class="template switches__group-label">Weakness</span>
        <div class="template switches__radiogroup">
          <input class="template switches__input switches__input--yes" name="switch-1" id="switch-1-yes" type="radio">
          <label class="template switches__radio-label switches__radio-label--yes" for="switch-1-yes">Yes</label>
          <input class="template switches__input switches__input--no" name="switch-1" id="switch-1-no" type="radio">
          <label class="template switches__radio-label switches__radio-label--no" for="switch-1-no">No</label>
          <div class="template switches__radiogroup-bg"></div>
        </div>
      </li>
      
      <li class="template switches__list-item">
        <span class="template switches__group-label">Neglect/Disregard/Extinction</span>
        <div class="template switches__radiogroup">
          <input class="template switches__input switches__input--yes" name="switch-2" id="switch-2-yes" type="radio">
          <label class="template switches__radio-label switches__radio-label--yes" for="switch-2-yes">Yes</label>
          <input class="template switches__input switches__input--no" name="switch-2" id="switch-2-no" type="radio">
          <label class="template switches__radio-label switches__radio-label--no" for="switch-2-no">No</label>
          <div class="template switches__radiogroup-bg"></div>
        </div>
      </li>
      
      <li class="template switches__list-item">
        <span class="template switches__group-label">Speech impairment</span>
        <div class="template switches__radiogroup">
          <input class="template switches__input switches__input--yes" name="switch-3" id="switch-3-yes" type="radio">
          <label class="template switches__radio-label switches__radio-label--yes" for="switch-3-yes">Yes</label>
          <input class="template switches__input switches__input--no" name="switch-3" id="switch-3-no" type="radio">
          <label class="template switches__radio-label switches__radio-label--no" for="switch-3-no">No</label>
          <div class="template switches__radiogroup-bg"></div>
        </div>
      </li>
      
    </ul>
    <hr>
    
  </main>
</body>

</html>"
`;

exports[`EjsAlgorithmRenderer Behavior should render text algorithm 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>ABCTs</title>
  <style>
    .template {
      box-sizing: border-box;
      margin: 0;
      padding: 0;
    }

    .template.switches__list,
    .template.outcomes__list {
      list-style: none;
      padding: 0;
    }

    .template.header__title {
      margin: 0;
      /* headline-medium */
      line-height: 36px;
      font-size: 28px;
      font-weight: 400;
    }

    .template.body__paragraph {
      margin-top: 16px;
      /* body-medium */
      line-height: 20px;
      font-size: 14px;
      letter-spacing: 0.25px;
      font-weight: 400;
    }

    .template.switches__list {
      margin-top: 16px;
    }

    .template.switches__list-item {
      display: flex;
      justify-content: space-between;
      align-items: center;
      padding-top: 8px;
      user-select: none;
    }

    .template.switches__list-item:first-child {
      padding-top: 0;
    }

    .template.switches__input {
      height: 0px;
      width: 0px;
      visibility: hidden;
    }

    .template.switches__group-label,
    .template.switches__radio-label {
      /* label-large */
      line-height: 20px;
      font-size: 14px;
      letter-spacing: 0.1px;
      font-weight: 500;
    }

    .template.switches__radiogroup {
      position: relative;
      padding: 4px 8px 4px 8px;
      border-radius: 999px;
      outline: 1px solid black;
      overflow: hidden;
    }

    .template.switches__radio-label {
      display: inline-block;
      width: 2em;
      text-align: center;
    }

    .template.switches__radiogroup-bg {
      background-color: black;
      position: absolute;
      top: 0;
      width: 50%;
      height: 100%;
      z-index: -1;
      display: none;
    }

    .template.switches__input:checked+.template.switches__radio-label {
      color: white;
    }

    .template.switches__input--yes:checked~.template.switches__radiogroup-bg {
      display: block;
      left: 0;
    }

    .template.switches__input--no:checked~.template.switches__radiogroup-bg {
      display: block;
      right: 0;
    }

    .template.outcomes__list-item {
      margin-top: 8px;
    }

    .template.outcomes__list-item:first-child {
      margin-top: 8px;
    }

    .template.outcomes__button {
      background-color: white;
      border: 1px solid black;
      border-radius: 999px;
      /* label-large */
      line-height: 20px;
      font-size: 14px;
      letter-spacing: 0.1px;
      font-weight: 500;
    }

    .template.outcomes__list-item {
      margin-top: 8px;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
    }

    .template.outcomes__list-item:first-child {
      margin-top: 8px;
    }

    .template.outcomes__button {
      background-color: black;
      color: white;
      border-radius: 999px;
      height: 40px;
      padding: 0 24px;
      /* label-large */
      line-height: 20px;
      font-size: 14px;
      letter-spacing: 0.1px;
      font-weight: 500;
    }
  </style>
</head>

<body>
  <header>
    <h1 class="template header__title">ABCTs</h1>
  </header>
  <main>
    <p class="template body__paragraph">
      ABCT = (airway, breathing, circulation, time of stroke onset)
    </p>
    <hr>
    
    <ul class="template outcomes__list">
      
      <li class="template outcomes__list-item">
        <button class="template outcomes__button">Stable</button>
      </li>
      
      <li class="template outcomes__list-item">
        <button class="template outcomes__button">Not Stable</button>
      </li>
      
    </ul>
    
  </main>
</body>

</html>"
`;
